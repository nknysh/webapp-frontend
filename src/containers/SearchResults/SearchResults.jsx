import React, { useState } from 'react';
import { compose, curry, length, map, path, defaultTo, prop } from 'ramda';

import { Card, Loader, Modal } from 'components';
import { SearchSidebar } from 'containers';
import { useCurrentWidth, useFetchData } from 'effects';
import { withSearchIndexes } from 'hoc';
import { isMobile, IndexTypes } from 'utils';

import uiConfig, { getPluralisation } from 'config/ui';

import connect from './SearchResults.state';
import { propTypes, defaultProps } from './SearchResults.props';
import {
  StyledResults,
  ResultsTitle,
  Results,
  Result,
  Filtering,
  FiltersButton,
  modalStyles,
} from './SearchResults.styles';

export const SearchResults = ({
  fetchSearch,
  getCountryName,
  getHotel,
  getResults,
  searchQuery,
  searchStatus,
  getHotelFeaturedPhoto,
}) => {
  const [modalOpen, setModalOpen] = useState(false);

  const loaded = useFetchData(searchStatus, fetchSearch);

  const currentWidth = useCurrentWidth();

  const id = path(['search', 'id'], searchQuery);
  const results = getResults(IndexTypes.HOTELS);

  const count = length(results) || 0;
  const destinationTitle = getCountryName(id) || '';
  const countTitle = `${count} ${getPluralisation('result', count)}`;
  const title = destinationTitle ? `${destinationTitle} - ${countTitle}` : countTitle;

  const onClose = () => setModalOpen(false);
  const onOpen = () => setModalOpen(true);

  const renderResult = curry((selector, hit) => {
    const id = prop('ref', hit);

    return (
      id && (
        <Result to={`/hotels/${id}`} key={id}>
          <Card {...selector(id)} featuredPhoto={getHotelFeaturedPhoto(id)} />
        </Result>
      )
    );
  });

  return (
    <Loader isLoading={!loaded} text={path(['messages', 'searching'], uiConfig)}>
      <StyledResults>
        <ResultsTitle>{title}</ResultsTitle>
        {isMobile(currentWidth) && (
          <Filtering>
            <FiltersButton onClick={onOpen}>{path(['buttons', 'refine'], uiConfig)}</FiltersButton>
            <Modal open={modalOpen} modalContentProps={modalStyles} onClose={onClose}>
              <SearchSidebar />
            </Modal>
          </Filtering>
        )}
        <Results>{map(renderResult(getHotel), defaultTo([], results))}</Results>
      </StyledResults>
    </Loader>
  );
};

SearchResults.propTypes = propTypes;
SearchResults.defaultProps = defaultProps;

export default compose(
  React.memo,
  withSearchIndexes([IndexTypes.HOTELS]),
  connect
)(SearchResults);
