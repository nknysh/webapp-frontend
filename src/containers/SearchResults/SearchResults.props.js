import PropTypes from 'prop-types';

import { noop } from 'utils';

export const propTypes = {
  searchQuery: PropTypes.object,
  indexes: PropTypes.array,
  hotels: PropTypes.object,
  destinations: PropTypes.object,
  getHotel: PropTypes.func,
  fetchDestinations: PropTypes.func,
  fetchHotels: PropTypes.func,
  getDestinationTitle: PropTypes.func,
};

export const defaultProps = {
  searchQuery: {},
  indexes: [],
  hotels: undefined,
  destinations: undefined,
  getHotel: noop,
  fetchDestinations: noop,
  fetchHotels: noop,
  getDestinationTitle: noop,
};
