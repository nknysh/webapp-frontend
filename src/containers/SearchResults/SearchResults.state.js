import { connect } from 'react-redux';
import { pipe } from 'ramda';

import { getHotelsFromSearchResults } from 'store/modules/hotels';
import {
  searchByQuery,
  getSearchStatus,
  getSearchQuery,
  getSearchResultsMeta,
  getSearchResultsResult,
  getCanSearch,
} from 'store/modules/search';

export const mapStateToProps = state => ({
  searchQuery: getSearchQuery(state),
  searchStatus: getSearchStatus(state, 'byQuery'),
  result: getHotelsFromSearchResults(state, getSearchResultsResult(state, 'byQuery')),
  meta: getSearchResultsMeta(state, 'byQuery'),
  canSearch: getCanSearch(state),
});

export const mapDispatchToProps = dispatch => ({
  searchByQuery: pipe(
    searchByQuery,
    dispatch
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
);
