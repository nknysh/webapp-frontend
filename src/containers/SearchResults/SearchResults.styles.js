import styled from 'styled-components';
import { Link } from 'react-router-dom';

import theme from 'styles/theme';
import breakpoints from 'styles/breakpoints';
import { Button } from 'styles/elements';
import { Heading2 } from 'styles/typography';

export const modalStyles = {
  style: { backgroundColor: theme.backgrounds.secondary, paddingTop: theme.gutter * 4 },
};

export const StyledResults = styled.div`
  padding ${theme.gutter * 2}px;

  ${breakpoints.tablet`
    padding: 0;
  `}
`;

export const ResultsTitle = styled(Heading2)`
  color: ${theme.neutral};
  display: block;
  padding: 0 0 ${theme.gutter}px;
  margin-top: ${theme.gutter}px;
  color: ${theme.neutral};
  text-transform: capitalize;
  font-weight: ${theme.fonts.bold};
  line-height: 29px;

  ${breakpoints.tablet`
    margin: ${theme.gutter}px ${theme.gutter / 2}px ${theme.gutter * 4}px;
  `}
`;

export const Results = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const Result = styled(Link)`
  cursor: pointer;
  width: 100%;

  ${breakpoints.tablet`
    flex: 0 1 50%;
    width: 50%;
  `}
`;

export const Filtering = styled.div`
  padding: ${theme.gutter}px 0;
  margin: 0 0 ${theme.gutter * 2}px;
  border-top: 1px solid ${theme.borders.default};
  border-bottom: 1px solid ${theme.borders.default};
`;

export const FiltersButton = styled(Button)`
  font-size: ${theme.fonts.sizes.normal}px;
`;
