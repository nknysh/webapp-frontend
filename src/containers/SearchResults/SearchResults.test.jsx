import React from 'react';

import { SearchResults } from './SearchResults';

const getComponent = props => shallow(<SearchResults getResults={() => []} {...props} />);

describe('<SearchResults />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
    });
  });
});
