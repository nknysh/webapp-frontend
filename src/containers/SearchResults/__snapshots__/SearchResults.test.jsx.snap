// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResults /> render matches snapshot 1`] = `
<Loader
  isLoading={true}
  showSpinner={true}
  text="Searching..."
>
  <styled.div>
    <Styled(styled.h2)>
      NaN properties found
    </Styled(styled.h2)>
    <styled.div />
  </styled.div>
</Loader>
`;
